File: MPC.m
clear;
clc;

N=5;    %N is the length of the model
P=5;   %P is the prediction horizon
M=3;    %M is the control horizon
r=1;

num = 1;
den = [5 1];

ts = 1; %Samplitng time


tr = tf(num,den);
sysd = c2d(tr, ts); 
s = step(sysd);
S=s(2:P+1);
%K = inv(Sf'*Sf)*Sf;
E = 1-s;
Sf = zeros(P,M);
Spast= zeros(P,N-2);

% for i = 0: P-1
%     Sf(P-i-2,(1:M-i)) = S(M-i:-1:1)
% end

for i = 1:P
    for j = 1:M
        if i-j+1 == 0
            break
        else
            Sf(i, j) = S(i-j+1);
        end
    end
end

File: MPC.py
import control as ctl
import matplotlib.pyplot as plt
import numpy as np

# Define the transfer function
num = [1]
den = [5, 1]
sys = ctl.TransferFunction(num, den)


N=5;    #N is the length of the model
P=5;    #P is the prediction horizon
M=3;    #M is the control horizon
r=1;
t_s = 1; #sampling time

num = 1;
den = [5, 1];


tr = ctl.TransferFunction(num,den);
sysd = ctl.c2d(tr, 1); 
td, rd = ctl.step_response(sysd);

rd = rd[1:N+1]

Sf = np.zeros((P,M))

for i in range(0,P):
    for j in range(0,M):
        if i-j < 0:
            break
        else:
            Sf[i][j] = rd[i-j]
